etcd3locker/locker.go

```diff
@@ -0,0 +1,120 @@
+// Tested on etcd 3.1/3.2/3.3
+package etcd3locker
+
+import (
+	"context"
+	"errors"
+	"log"
+	"strings"
+	"sync"
+
+	etcd3 "github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/tus/tusd"
+)
+
+var (
+	ErrLockNotHeld = errors.New("Lock not held")
+	DefaultTtl     = int64(60)
+)
+
+type Etcd3Locker struct {
+	// etcd3 client session
+	Client *etcd3.Client
+
+	// locks is used for storing Etcd3Locks before they are
+	// unlocked. If you want to release a lock, you need the same locker
+	// instance and therefore we need to save them temporarily.
+	locks  map[string]*Etcd3Lock
+	mutex  sync.Mutex
+	prefix string
+}
+
+// New constructs a new locker using the provided client.
+func New(client *etcd3.Client) (*Etcd3Locker, error) {
+	return NewWithPrefix(client, "/tusd")
+}
+
+// This method may be used if a different prefix is required for multi-tenant etcd clusters
+func NewWithPrefix(client *etcd3.Client, prefix string) (*Etcd3Locker, error) {
+	locksMap := map[string]*Etcd3Lock{}
+
+	if !strings.HasPrefix(prefix, "/") {
+		prefix = "/" + prefix
+	}
+
+	return &Etcd3Locker{Client: client, prefix: prefix, locks: locksMap, mutex: sync.Mutex{}}, nil
+}
+
+// UseIn adds this locker to the passed composer.
+func (locker *Etcd3Locker) UseIn(composer *tusd.StoreComposer) {
+	composer.UseLocker(locker)
+}
+
+// LockUpload tries to obtain the exclusive lock.
+func (locker *Etcd3Locker) LockUpload(id string) error {
+	session, err := locker.createSession()
```

Comment:
Is it necessary to create a new session for every lock or could they be reused?

Reply:
The locks don't work if cleared/re-created in the same session - my observations led me to conclude that if you're the same "session owner" re-establishing the lock - it'll clear it.

Reply:
> if you're the same "session owner" re-establishing the lock - it'll clear it.

Ok, that's not the behavior we want, thanks for clarifying.